plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java

```diff
@@ -69,68 +78,100 @@ public RepositoryStatistics mine(final Collection<String> absoluteFileNames, fin
-                        logger.logError("Can't obtain HEAD of repository.");
-                        return statistics;
-                    }
-                    Set<String> files = new FilesCollector(repository).findAllFor(head);
-                    return analyze(repository, files);
+                try (Git git = new Git(repository)) {
+                    List<RevCommit> commits = new CommitCollector(repository, git).findAllCommits();
+                    return analyze(repository, git, commits);
+                }
+                catch (GitAPIException | IOException exception) {
+                    logger.logException(exception, "Can't obtain all commits for the repository.");
                 }
-                return analyze(repository, paths);
-            }
-            catch (IOException exception) {
-                RepositoryStatistics statistics = new RepositoryStatistics();
-                logger.logException(exception, "Can't obtain HEAD of repository.");
-                return statistics;
             }
             finally {
                 repository.close();
             }
+
+            return new RepositoryStatistics();
         }
 
-        RepositoryStatistics analyze(final Repository repository, final Collection<String> files) {
+        RepositoryStatistics analyze(final Repository repository, final Git git, final List<RevCommit> commits)
+                throws IOException {
             RepositoryStatistics statistics = new RepositoryStatistics();
-            logger.logInfo("Invoking Git miner to create statistics for all available files");
-            logger.logInfo("Git working tree = '%s'", getWorkTree(repository));
-
             FileStatisticsBuilder builder = new FileStatisticsBuilder();
-            List<FileStatistics> fileStatistics = files.stream()
-                    .map(builder::build)
-                    .map(file -> analyzeHistory(repository, file))
-                    .collect(Collectors.toList());
-            statistics.addAll(fileStatistics);
-
-            logger.logInfo("-> created statistics for %d files", statistics.size());
-
+            List<String> files = new ArrayList<>();
+            Map<String, FileStatistics> fileStatistics = new HashMap<>();
+            Set<String> filesInHead = new FilesCollector(repository).findAllFor(repository.resolve(Constants.HEAD));
+            for (int i = commits.size() - 1; i >= 0; i--) {
+                if (i == commits.size() - 1) {
+                    files = getFilesFromCommit(repository, git, null, commits.get(i).getName());
+                }
+                else {
+                    if (i > 0) {
+                        files = getFilesFromCommit(repository, git, commits.get(i + 1).getName(),
+                                commits.get(i).getName());
+                    }
+                }
+                int finalI = i;
+                files.forEach(f -> fileStatistics.computeIfAbsent(f, builder::build)
```

Comment:
Wouldn't it be simpler to extract the files from the diff?

Reply:
I'm not sure if I understand this point. I am extracting the file paths in ``getFilesFromCommit`` from the diff and create the corresponding FileStatistics here.

Reply:
Sorry,  I thought you are iterating over `filesInHead`.